tutorials/Land/Soil/Coupled/equilibrium_test.jl

```diff
@@ -0,0 +1,556 @@
+# The thermal conductivity of air is a constant parameter used across ClimateMachine,
+# so we will get it from that parameter set.
+κ_air = FT(K_therm(param_set)); # W/m/K
+# The particle density of organic material-free soil is
+# equal to the particle density of quartz and other minerals [3]:
+ρp = FT(2700); # kg/m^3
+# We calculate the thermal conductivities for the solid material, for dry soil,
+# and for saturated soil. These functions are taken from Balland and Arp (2005) [3].
+κ_solid = k_solid(ν_ss_om, ν_ss_quartz, κ_quartz, κ_minerals, κ_om)
+κ_dry = k_dry(κ_solid, porosity, κ_air, ρp)
+κ_sat_frozen = ksat_frozen(κ_solid, porosity, κ_ice)
+κ_sat_unfrozen = ksat_unfrozen(κ_solid, porosity, κ_liq);
+# Next, we calculate the volumetric heat capacity of dry soil. Dry soil
+# refers to soil that has no water content.
+ρc_ds = FT((1 - porosity) * 1.926e06); # J/m^3/K
+# We collect the majority of the parameters needed
+# for modeling heat and water flow in soil in `soil_param_functions`.
+# The van Genuchten parameters are stored in the water model, in an object
+# called `hydraulics`, however.
+
+soil_param_functions = SoilParamFunctions{FT}(
+    Ksat = Ksat,
+    S_s = S_s,
+    porosity = porosity,
+    ν_ss_gravel = ν_ss_gravel,
+    ν_ss_om = ν_ss_om,
+    ν_ss_quartz = ν_ss_quartz,
+    ρc_ds = ρc_ds,
+    κ_dry = κ_dry,
+    κ_sat_unfrozen = κ_sat_unfrozen,
+    κ_sat_frozen = κ_sat_frozen,
+    a = 0.24, # used in the Kersten number [2]
+    b = 18.1, # used in the Kersten number [2]
+);
+
+
+# # Initial and Boundary conditions
+# Dirichlet boundary conditions are on `ϑ_l` and
+# `T`, while Neumann boundary conditions are on `-κ∇T` and `-K∇h`. For Neumann
+# conditions, the user supplies a scalar, which is multiplied by `ẑ` within the code.
+
+# The initial water profile:
+function ϑ_l0(aux)
+    FT = eltype(aux)
+    zmax = FT(0)
+    zmin = FT(-1)
+    theta_max = FT(porosity * 0.5)
+    theta_min = FT(porosity * 0.4)
+    c = FT(20.0)
+    z = aux.z
+    output =
+        theta_min +
+        (theta_max - theta_min) * exp(-(z - zmax) / (zmin - zmax) * c)
+    return output
+end;
+
+
+# Water boundary conditions:
+surface_water_flux = (aux, t) -> eltype(aux)(0.0)
+bottom_water_flux = (aux, t) -> eltype(aux)(0.0)
+surface_water_state = nothing
+bottom_water_state = nothing;
+
+
+# The initial temperature profile:
+function T_init(aux)
+    FT = eltype(aux)
+    zmax = FT(0)
+    zmin = FT(-1)
+    T_max = FT(289.0)
+    T_min = FT(288.0)
+    c = FT(20.0)
+    z = aux.z
+    output = T_min + (T_max - T_min) * exp(-(z - zmax) / (zmin - zmax) * c)
+    return output
+end;
+
+
+# As we are not including the equations for phase changes in this tutorial,
+# we chose temperatures that are above the freezing point of water.
+
+# The boundary conditions for the heat equation:
+surface_heat_flux = (aux, t) -> eltype(aux)(0.0)
+bottom_heat_flux = (aux, t) -> eltype(aux)(0.0)
+surface_heat_state = nothing
+bottom_heat_state = nothing;
```

Comment:
What does a `nothing` surface state imply?

Reply:
It means that no BC is applied to that surface. We have four possibilities per eqn (dirichlet top, bottom, neumann top, bottom) but should only specify two. In the BC functions, we check, and if the BC is not nothing, we apply the BC supplied

Reply:
Does no BC mean no flux? or 0 Dirichlet? I think avoiding `nothing` for BC application is more clear. I would just suggest not including the BC, and letting the constructor handle default BCs (whether they're Dirichlet / Neumann)

Reply:
no BC means no BC (no flux and 0 dirichlet are BCs). The idea is that since the Eqns are 2nd order in spatial derivatives, we need to supply two pieces of information (which could be zero flux of zero state) to uniquely determine the solution. But, if we restrict to Neumann and Dirichlet BC types, there are four possible pieces of information to supply. so, the two that arent picked are "no BC" or `nothing`.

I think this defn: https://github.com/CliMA/ClimateMachine.jl/blob/4607e850cbe7d33247cfd5141d7907c9afdafaac/src/Land/Model/land_bc.jl#L40

allows us to do what you suggested though - For example, we can instead call the soil model with:
```soil_water_model = SoilWaterModel(
           FT;
           viscosity_factor = TemperatureDependentViscosity{FT}(),
           moisture_factor = MoistureDependent{FT}(),
           hydraulics = vanGenuchten{FT}(α = vg_α, n = vg_n),
           initialϑ_l = ϑ_l0,
           dirichlet_bc = Dirichlet(),
           neumann_bc = Neumann(
               surface_flux = surface_water_flux,
               bottom_flux = bottom_water_flux,
           ),)```

and this works the same as specifying "nothing" for both the Dirichlet options.

What do you think?

Reply:
> no BC means no BC

I think we need to see what the default behavior is. Is this equivalent to leaving the `boundary_state!` kernel empty? @jkozdon would this be a zero Neumann BC? I'm not sure I understand what "no BCs" mean.

Reply:
Here is an example of how it is used in the `boundary_state!` function:
https://github.com/CliMA/ClimateMachine.jl/blob/25bc89432e183fd53b96afe0cfd09b5165407eb4/src/Land/Model/soil_bc.jl#L501

Reply:
> no BC means no BC (no flux and 0 dirichlet are BCs). The idea is that since the Eqns are 2nd order in spatial derivatives, we need to supply two pieces of information (which could be zero flux of zero state) to uniquely determine the solution. But, if we restrict to Neumann and Dirichlet BC types, there are four possible pieces of information to supply. so, the two that arent picked are "no BC" or `nothing`.
> 
> I think the lines following this:
> 
> https://github.com/CliMA/ClimateMachine.jl/blob/4607e850cbe7d33247cfd5141d7907c9afdafaac/src/Land/Model/land_bc.jl#L40
> 
> allow us to do what you suggested though - For example, we can instead call the soil model with:
> 
> ```
>            FT;
>            viscosity_factor = TemperatureDependentViscosity{FT}(),
>            moisture_factor = MoistureDependent{FT}(),
>            hydraulics = vanGenuchten{FT}(α = vg_α, n = vg_n),
>            initialϑ_l = ϑ_l0,
>            dirichlet_bc = Dirichlet(),
>            neumann_bc = Neumann(
>                surface_flux = surface_water_flux,
>                bottom_flux = bottom_water_flux,
>            ),)
> ```
> 
> and this works the same as specifying "nothing" for both the Dirichlet options.
> 
> What do you think? Just a driver level change?

to clarify, the four possible options arise because this is 1d - so we have (top, bottom) x (dirichlet, neumann) as our options.

Reply:
OK - I added this text to the tutorials. https://github.com/CliMA/ClimateMachine.jl/blob/bdf33a47165f258a7eb78535941f4ca12e06306d/tutorials/Land/Soil/Coupled/equilibrium_test.jl#L230

I think the mechanics of the code is OK - as a user, it is natural that I have to give two BC. The code has the flexibility to accept any combination (Dirichlet, Neumann, top, bottom), so setting `nothing` for the fields where I dont want to give a BC feels reasonable.  What "do nothing" in the `boundary_state!` function is another question that the user shouldnt have to care about

Whether or not the two BC supplied must be at the top and bottom is another question (I put in the tutorial text that they need to be). So unless we want to add in a check that requires them to be at the top and at the bottom, not both at the top, e.g., this seems fine to me now.